keps/sig-apps/2214-array-job/README.md

```diff
@@ -0,0 +1,572 @@
+- [x] (R) Graduation criteria is in place
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [x] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP extends kubernetes with user-friendly support for running
+embarrassingly parallel jobs.
+
+Here, parallel means multiple pods. By embarrassingly parallel, it means that
+the pods have no dependencies between each other.
+In particular, neither ordering between pods nor gang scheduling are supported.
+
+We propose the addition of completion indexes into the Pods of a *Job
+[with fixed completion count]* to support running embarrassingly parallel
+programs, with a focus on ease of use.
+This new Job pattern is called an *array Job*, because each Pod of the Job is
+specialized to work on a particular index, as if the Pods where elements of an
+array.
+
+[with fixed completion count]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs
+
+## Motivation
+
+Users can use some [Job patterns] to run embarrassingly parallel Jobs, but those
+approaches have downsides:
+
+- The queue patterns require setting up an external queue service and modifying
+  the Job binary to be able to connect to the queue.
+  Depending on the implementation, it is prone to race conditions when
+  coordinating which Pod works on which item.
+- The template pattern doesn't scale when the parallelism level is too high,
+  in terms of job creation and querying status.
+
+Due to these reasons, workloads where each Pod just needs a unique and
+ordered completion index, are hard to adapt to the existing Job patterns.
+
+The lack of support for this pattern in k8s forces users to implement their
+own APIs and controllers or adopt third party implementations. Each
+implementation splits the ecosystem, making it harder for higher level systems
+for Job queueing or workflows to support all of them.
+
+[Job patterns]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-patterns
+
+### Goals
+
+- Support the *array Job* pattern by adding completion indexes to each Pod
+  of a Job in *fixed completion count* mode.
+- Track the completion for each Job Pod per completion index.
+- Handle varying parallelism by creating Pods with lower indexes first and
+  removing Pods with higher indexes first. Jobs that do not require completion
+  indexes preserve their downscaling behavior.
```

Comment:
This already described the solution, not just what we want to achieve.

Reply:
so just: "Handle varying parallelism"

Reply:
I think the first goal alone is sufficient, the rest is implementation details.

Reply:
Removed